streamr-dev/streamr-platform
# Redirect to login after 403

If unauthenticated and hitting a 403 on loading a canvas/dashboard, redirect to login with redirect set to original canvas/dashboard. If logged in, redirect to /404. i.e. don't ask logged in users to log in again if they don't have access to something.

Also:

* Fixes link copied when clicking "copy link" in share dialog.
* Fixes live data flowing on public/embed canvases (bad merge 👎).

Let's do a deploy if/when this PR lands.

## To Test

### 1. Create Running Canvas, Copy link
1. Create Canvas
2. Connect a Clock to a Table
3. Start canvas
4. Open Share dialog, hit "Copy Link".

### 2. Test 404 works for logged in
1. Open new Tab, paste url in new tab
3. Delete last char from url (i.e. it's invalid), press enter.
4. It should redirect you to /404

### 2. Test 404 works for logged out
1. Open incognito window, paste url
2. Delete last char from url (i.e. it's invalid), press enter.
3. It should redirect you to /404

### 3. Test redirect works for logged out
1. Paste full url in incognito address bar again, press enter.
2. It should redirect you to login
3. Log in as the same user
4. It should redirect you to the running canvas.

### 4. Test public canvas data flows
1. Close incognito window.
2. Open Share dialog, make canvas public.
3. Paste canvas url in incognito window again.
4. Canvas should open and data should be flowing even though you're logged out.
5. Log in, go to same url.
6. Canvas should open and data should be flowing
